Test/add test for bonding curve contract setup

0xNuggan opened the pull request:
Tests for the Bonding Curve contracts. Notes:

- I think I found a logic issue in _convertAmountToRequiredDecimal and provided an alternative. I may be missing something, please review. @Zitzak

- The formula starts reverting when supplying some numbers that don't seem that high to me, specially when selling (assuming 18 decimals). I still want to look closer into it, but extra eyes can only help.

- Since there is some rounding and precision loss involved, some test use foundry 'assertApproxEqRel' instead of 'assertEq'.

- The 4 functions in the coverage of the Bancor funding manager are the stubs for the IFundingManager interface. We should think about removing those functions from the interface and only leave transferOrchestratorToken().

- There are a couple of other bugfixes and changes, usually pretty straightforward. They are also tagged with the @review tag

Zitzak committed:
feat(BondingCurveFundingManagerBase): Add base features

The base contract enables the basic functionalitiy of issuing token
which every bonding curve must have. It makes use of a abstract function
for the calculation which should be implemented in downstream contracts

Zitzak committed:
feat(RedeemingBondingCurveFundingManager): Add redeem functionalities

This base contract adds configuration function for a redeeming bonding
curve functionality as well as abastract function for calculating
token values to be implemented in an downstream contract

Zitzak committed:
feat(VirtualSupply): implements virtual token and collateral supply

Zitzak committed:
feat(BancorVirtualSupplyBondingCurveFundingManger):

Inherits from the base contracts to enable issuing and redeeming
of tokens on a bonding curve, calculated through the Bancor formula,
based on virtual supplies of the issuing token and collateral

Zitzak committed:
feat:BancorVirtualSupplyBondingCurveFundingManager

Remove Virtual Supply token and collateral interfaces

Zitzak committed:
feat: add getter func and errors to interfaces

The getter function got added as per PR review request.  Next to this
errors got added to the interface for over/underflow when setting
the virtual supply

Zitzak committed:
feat: add setter func to VirtualSupply contract

The setter function has been implemented as abstract
function so the upstream contract has to implement it.
Because the getter functions got added to the interfaces,
both contract now implement the interface. Lastly some
input validation got added

Zitzak committed:
test: virtual supply token contracts

Zitzak committed:
feat: add Bancor contracts to repo

All contracts have added comments for clarification
what has been changed, including links to the
original files.

An ignore for the formula directory got added to the
foundry.toml file to prevent the formatter from creating
to big of a diff, in case somebody would like to check our
statement of the changes made.

Zitzak committed:
chore: update modifier after rebase

Zitzak committed:
chore: remove collateral token address

The fundingManager infrastructure currently only works with one tokens.
For that reason it is unneeded to facilitate for
multi collateral token possibility, rather build for what is
required now.

Zitzak committed:
chore: WIP

0xNuggan committed:
Add VirtualToken/Collateralbase tests

0xNuggan committed:
Merge branch 'main' into test/add_test_for_bondingCurve_contract_setup

0xNuggan committed:
address merge issues

0xNuggan committed:
set up testing skeleton

0xNuggan committed:
BondingCurveManagerBase Test + small bugfixes

bugfixes in the contracts are marked with @review tag

0xNuggan committed:
RedeemingBondingCurve test skeleton + minor cleanup

0xNuggan committed:
small corrections

expands the redeemingFundingManager, some formatting and implements a stopgap solution for stackTooDeep error

0xNuggan committed:
bump Base Bonding Curve test coverage to 100%

0xNuggan committed:
Rename test file

It was messing up the coverage table (ò_ó)

0xNuggan committed:
finish RedeemingBC tests + bugfixes

0xNuggan committed:
bump coverage to 100% in RedeemBC

0xNuggan committed:
WIP Bancor tests

0xNuggan committed:
format + fix typo

0xNuggan committed:
WIP BancorBondingCurve Tests

0xNuggan committed:
format + substitute msg.sender

0xNuggan committed:
decimalConversion tests + bugfix

0xNuggan committed:
bancor tests v1 ready

0xNuggan committed:
add scripts + e2e and bump coverage

Zitzak committed:
chore: review bugfixes and add getter for Bancor Funding Manager

What has been done?
- Reviewd bug fixes and removed the comments
- Add getter functions as suggested by comment
- Add tests for getter functions

Zitzak committed:
chore: remove comment of idempotence

Why?
Thought process here is that it saves gas to not needlessly execute
a tx. If the PR reviewers would like a different practice then
I'm happy to change the code, because it might be personal pref

Zitzak committed:
chore: remove bugfix comments of Redeeming and Base BCFM

Reviewd all the comments and agreed. Removed the comments

Zitzak committed:
docs: add documentation for max computational amount Bancor BCFM

The Bancor Formula contract has a maximum uint value it can process
for issuing and redeeming, 10^26 and 10^38 respectively. When a
larger value is used for calculations, the contract will revert.

These values leave enough room for most cases but the edge cases.
Edge cases could include tokens that have very little value or
tokens with high decimals. For those cases it might be necessary
to split buy amounts over multiple transaction.

Because we don't want to modify the Aragon formula contract we decided to
make sure to inform the user and dev of this fact by means of a
natspec documentation

Zitzak committed:
WIP

Zitzak committed:
feat: add events to bonding curve funding manager

Adds all the events and tests for:
- BondingCurveFundingManagerBase
- RedeemingBondingCurveFundingMangerBase
- VirtualCollateralSupplyBase
- VirtualTokenSupplyBase
- BancorVirtualSupplyBondingCurveFundingManager

0xNuggan committed:
emit bugfix

0xNuggan committed:
Update RedeemingB_CurveFundingManagerBase.t.sol

0xNuggan committed:
Update BondingCurveFundingManagerBase.t.sol

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BondingCurveFundingManagerBase.sol:
```diff
@@ -81,63 +105,132 @@ abstract contract BondingCurveFundingManagerBase is
+            revert BondingCurveFundingManager__InvalidDepositAmount();
+        }
+        // Transfer collateral, confirming that correct amount == allowance
+        __Module_orchestrator.token().safeTransferFrom(
```
__Module_orchestrator.token() wont exist  in the future.
It might make sense to reference a token specified in the BondingCurveFundingManagerBase

Zitzak replied:
Do we already know when this change will be implemented? This would mean that until then the module would not be able to work with the current setup.

Also seeing that all `FundingManagers` will need to have a token specified, might it make sense to adapt the base `Module` and/or `IFundingmanager` contract/interface?

FHieser replied:
IFundingManager has already the token() function implemented and can just be referenced via orchestrator.fundingManager.token() i think

Zitzak replied:
Will implement the IFundingManager interface as you suggested as well as the way to reference the token :) thanks

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BondingCurveFundingManagerBase.sol:
```diff
@@ -20,58 +20,82 @@ import {ERC20Upgradeable} from "@oz-up/token/ERC20/ERC20Upgradeable.sol";
+///             the downstream contract.
 abstract contract BondingCurveFundingManagerBase is
     IBondingCurveFundingManagerBase,
-    IFundingManager,
```
It says FundingManagerBase in the title,
Why is this not a Ifundingmanager?

0xNuggan replied:
This ties to the comment above. Probably a good call, if BondingCurveFundingManagerBase becomes a IFundingManager, it can call self.token() and avoid getting it through the proposal, which will be deprecated anyway. What do you think @Zitzak  ?

Zitzak replied:
I think it makes sense indeed.

The redeeming contract however also has FundingManagerBase in the title. Would you suggest renaming it? @FHieser

0xNuggan replied:
I think since it inherits from bondingCurvefundingManagerBase (and therefore inherits IFundingManager) it's ok to keep the name as is

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BondingCurveFundingManagerBase.sol:
```diff
@@ -81,63 +105,132 @@ abstract contract BondingCurveFundingManagerBase is
+    /// This function updates the `tokenDecimals` state variable.
+    /// @param _decimals The number of decimals to set for the token.
     function _setTokenDecimals(uint8 _decimals) internal {
-        // Input check
```
Does this not need an input check?
Check for zero
Check / Reference possible Formular borders

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/IBancorVirtualSupplyBondingCurveFundingManager.sol:
```diff
@@ -0,0 +1,43 @@
+    //--------------------------------------------------------------------------
+    // Events
+
+    // TODO: Contract still needs events. Where are events needed?
```
TODO

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/RedeemingBondingCurveFundingManagerBase.sol:
```diff
@@ -2,70 +2,193 @@
+            revert RedeemingBondingCurveFundingManager__InvalidDepositAmount();
+        }
+        // Transfer issued token to contract, confirming deposit amount == allowance
+        IERC20(address(this)).safeTransferFrom(
```
Does this work?
Isnt this an internal call?

FHieser replied:
Also just out of curiosity:
Shouldnt the tokens be burned here?

0xNuggan replied:
If there's enough allowance, it works. To your second point: you're absolutely right, we should burn there. The virtual balances are accounted for in the Bancor tests, but the actual on-chain one slipped through the cracks. :facepalm:

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/marketMaker/IVirtualTokenSupply.sol:
```diff
@@ -3,9 +3,31 @@
+    //--------------------------------------------------------------------------
+    // Events
+
+    // TODO: Contract still needs events. Where are events needed?
```
TODO

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/marketMaker/VirtualCollateralSupplyBase.sol:
```diff
@@ -4,36 +4,66 @@ pragma solidity 0.8.19;
+    /// @dev The internal state variable to keep track of the virtual collateral supply.
+    uint internal virtualCollateralSupply;
+    /// @dev Maximum unsigned integer value for overflow checks.
+    uint private constant MAX_UINT = type(uint).max; // TODO: The maximum supply has to be divided by the BPS precision used. If not, it can come to overflow errors in downstream contracts
```
TODO

Zitzak replied:
Could you comment what was meant by this @0xNuggan. I was unsure what to do here

0xNuggan replied:
Old comment. I removed it now

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/marketMaker/VirtualCollateralSupplyBase.sol:
```diff
@@ -4,36 +4,66 @@ pragma solidity 0.8.19;
+    // Public Functions Implemented in Downstream Contract
+
+    /// @inheritdoc IVirtualCollateralSupply
+    function setVirtualCollateralSupply(uint _virtualSupply) external virtual;
```
why is this virtual?
Is this coupled to a downstream contract and if so maybe an explanation in the docs is in order

0xNuggan replied:
I'd say in case a downstream contract wants to set specific limitations? It's not supposed to be able to work by itself anyway...

Zitzak replied:
The reason is that the implementing function needs to be restricted by a modifier, which cannot be accessed in this contract. For that reason this function is set to abstract.

I'll add a comment to clarify this

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/marketMaker/VirtualTokenSupplyBase.sol:
```diff
@@ -4,35 +4,66 @@ pragma solidity 0.8.19;
     uint internal virtualTokenSupply;
-    uint private constant MAX_UINT = 2 ** 256 - 1;
+    /// @dev Maximum unsigned integer value for overflow checks.
+    uint private constant MAX_UINT = type(uint).max; // TODO: The maximum supply has to be divided by the BPS precision used. If not, it can come to overflow errors in downstream contracts
```
TODO

Zitzak replied:
Same as the other one :)

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/marketMaker/VirtualTokenSupplyBase.sol:
```diff
@@ -4,35 +4,66 @@ pragma solidity 0.8.19;
+    // Public Functions Implemented in Downstream Contract
+
+    /// @inheritdoc IVirtualTokenSupply
+    function setVirtualTokenSupply(uint _virtualSupply) external virtual;
```
See virtual collateral Token supply

0xNuggan replied:
same as above

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.sol:
```diff
@@ -55,104 +94,386 @@ contract BancorVirtualSupplyBondingCurveFundingManager is
-        uint amountIssued = _issueTokens(_depositAmount, collateral);
-        _addTokenAmount(amountIssued);
-        _addCollateralAmount(_depositAmount);
+        _virtualSupplyBuyOrder(_msgSender(), _depositAmount);
```
hm maybe go with:
```
super().buyOrder(...);
_virtualSupplyBuyOrder(...);
```
out of the name _virtualSupplyBuyOrder i would just assume that it creates a buyorder for the virtualsupply only

0xNuggan replied:
Not sure about this one... It does create a buy order with the virtual supply, and not the actual one. I created the function to avoid code duplication. Of course we can also make _buyOrder virtual and override it, but to me this seems more explicit?

0xNuggan replied:
maybe renaming to virtualBuyOrder is better ?

FHieser replied:
I was a bit confused, when i read it, because you combine the two of them together basically, but the name suggests that you use just the virtual

Zitzak replied:
I think the idea is that the base contract has the `buyOrder` function of course. But because our implementation makes use of a virtual supply, we prepend the virtual supply part to the buyOrder.

I think overriding _buyOrder would defeat the purpose of our abstract formula wrappers

FHieser replied:
But arent you overriding it currently anyway?
I think for me it was more of a design thing anyway.
Because your using the original _buyOrder function anyway this doesnt make so much of a difference,
but I thought because you extend it anyway you could also write it the way I described it before

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.sol:
```diff
@@ -55,104 +94,386 @@ contract BancorVirtualSupplyBondingCurveFundingManager is
+        validReceiver(_receiver)
+        sellingIsEnabled
+    {
+        _virtualSupplySellOrder(_receiver, _depositAmount);
```
See comment above

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.sol:
```diff
@@ -55,104 +94,386 @@ contract BancorVirtualSupplyBondingCurveFundingManager is
-        uint redeemAmount = _redeemTokens(_depositAmount, collateral);
-        _subTokenAmount(_depositAmount);
-        _subCollateralAmount(redeemAmount);
+        _virtualSupplySellOrder(_msgSender(), _depositAmount);
```
same here

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.sol:
```diff
@@ -55,104 +94,386 @@ contract BancorVirtualSupplyBondingCurveFundingManager is
+    ///
+    /// @param _reserveRatio The reserve ratio to be set for buying tokens. Must be <= PPM.
+    function _setReserveRatioForBuying(uint32 _reserveRatio) internal {
+        // TODO: Qs - TEST: What happens when set to 0? -> Reserve ratio of 0 is not allowed
```
TODO

0xNuggan replied:
Yep, forgot to remove that one

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.sol:
```diff
@@ -55,104 +94,386 @@ contract BancorVirtualSupplyBondingCurveFundingManager is
+    ///
+    /// @param _reserveRatio The reserve ratio to be set for selling tokens. Must be <= PPM.
+    function _setReserveRatioForSelling(uint32 _reserveRatio) internal {
+        // TODO: Qs - TEST: What happens when set to 0? -> Reserve ratio of 0 is not allowed
```
TODO

0xNuggan replied:
same as above

FHieser commented on test/e2e/E2eTest.sol:
```diff
@@ -87,6 +91,43 @@ contract E2eTest is Test {
+        "BancorVirtualSupplyBondingCurveFundingManager"
+    );
+
+    function setUpBondingCurveFundingManager() private {
```
setUpBancorVirtualSupplyBondingCurveFundingManager

0xNuggan replied:
Supercalifragilisticexpialidocious

FHieser replied:
xD
At some point this probably isnt the only one that we use lol

FHieser commented on test/modules/fundingManager/bondingCurveFundingManager/marketMaker/utils/mocks/BancorVirtualSupplyBondingCurveFundingManagerMock.sol:
```diff
@@ -0,0 +1,54 @@
+    BancorVirtualSupplyBondingCurveFundingManager
+{
+    //--------------------------------------------------------------------------
+    // The BancorVirtualSupplyBondingCurveFundingManager is not abstract, so all the necessary functions are already implemented
```
Just FYI:
Instead of creating a mock here to test internal functions, sometimes you can also just inherit from the contract in the test file and that way you can call it internally

0xNuggan replied:
Yeah, I know, but it always seems confusing to me... I prefer to instantiate mocks and test them. But if we choose to enforce a specific code standard I'm happy to oblige.

FHieser commented on test/e2e/BondingCurve.t.sol:
```diff
@@ -0,0 +1,138 @@
+        // Note that we simulate orchestrator spending by just burning tokens.
+        uint halfOfDeposit = token.balanceOf(address(fundingManager)) / 2;
+        fundingManager.setVirtualCollateralSupply(halfOfDeposit);
+        /*         token.burn(
```
outcommented the burning of the tokens

0xNuggan replied:
Forgot to remove the comment. Since we are using a virtual supply we don't need to actually burn anything.

FHieser commented on test/modules/fundingManager/bondingCurveFundingManager/RedeemingB_CurveFundingManagerBase.t.sol:
```diff
@@ -0,0 +1,461 @@
+    }
+
+    /* Test sellOrder and _sellOrder function
+        ├── when the sell amount is 0
```
@0xNuggan did you write this tree yourself?

0xNuggan replied:
Yes. But apparently missed one statement to be controlled for...

FHieser commented on test/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.t.sol:
```diff
@@ -0,0 +1,1036 @@
+
+    are not tested since they are empty and will be removed in the future.
+
+    Also, since the following functions just wrap the formula contract, their content is assumed to be tested in the formula tests, not here:
```
Currently there arent any formular tests

0xNuggan replied:
Because we are taking the formula wholesale from Aragon, so we don't re-test it. Updated the wording

FHieser reviewed (commented):
Structure wise it looks good
This was more of a general review and things i noticed while looking broadly over it

0xNuggan committed:
review fixes

review fixes

- make base bonding curve FundingManager
-rename virtualBuy/Sell orders
- fix burn bug
- add check for edge case in testing
- other smaller stuff

0xNuggan committed:
Merge branch 'test/add_test_for_bondingCurve_contract_setup' of github.com:InverterNetwork/inverter-contracts into test/add_test_for_bondingCurve_contract_setup

Zitzak committed:
fix: add decimal input validation

Zitzak committed:
docs: add comment clarifying abstract functions

Zitzak committed:
chore: add events and remove todo comments

0xNuggan committed:
remove todo

0xNuggan committed:
Merge branch 'main' into test/add_test_for_bondingCurve_contract_setup

0xNuggan committed:
resolve merge issues

0xNuggan committed:
make fmt

0xNuggan committed:
address edge case in ElasticReceiptToken

FHieser reviewed (approved):
Although Im giving the approve I would highly suggest we get another pair on eyes on this, because of the volume of this PR

0xNuggan committed:
Merge branch 'dev' into test/add_test_for_bondingCurve_contract_setup
